pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,509 @@
+        }
+
+        template<class T>
+        void fput_multiple(const T * src, const uint64_t n_elements) {
+            _check_put();
+            _mmap_w->fput_multiple<T>(src, n_elements);
+        }
+
+    private:
+        enum class _Mode {
+            UNINIT = 0,             // Uninitialized
+            READONLY = 1,           // "r"
+            WRITEONLY = 2,          // "w"
+        };
+
+        _details::_MmapFileSave * _mmap_w = nullptr; // Used for write mode
+        _details::_MmapFileLoad * _mmap_r = nullptr; // Used for read mode
+        _Mode _mode = _Mode::UNINIT; // Indicator for current opened mode
+
+        /* Check if opened as read mode for calling get functions */
+        void _check_get() {
+            if (!_mmap_r) {
+                throw std::runtime_error("Not opened for read mode, cannot call get.");
+            }
+        }
+
+        /* Check if opened as write mode for calling put functions */
+        void _check_put() {
+            if (!_mmap_w) {
+                throw std::runtime_error("Not opened for write mode, cannot call put.");
+            }
+        }
+}; // end class MmapFile
+
+
+/* Extended Vector class that can be used as a std::vector or as a memory-view from part of mmap file
+ * For std::vector case, it own the memory for data storage.
+ * For mmap view case, it does not own any memory, but serve as a view for a piece of memory owned by MmapFile.
+ * By default, it is initialized as empty std::vector that can be resized or loaded as mmap view.
+ * Once loaded as mmap view, it cannot go back to std::vector case unless clear() is called.
+ */
+template<class T, class TT = T, _details::if_simple_serializable<TT> = true>
+class MmapableVector {
+    public:
+        MmapableVector(uint64_t size=0) { resize(size); }
+
+        /* Functions to match std::vector interface */
+        uint64_t size() const { return _size; }
+
+        const T* data() const { return _data; }
+
+        T& operator[](uint64_t idx) { return _data[idx]; }
+        const T& operator[](uint64_t idx) const { return _data[idx]; }
+
+        bool empty() const { return static_cast<bool> (_size == 0); }
+
+        void resize(uint64_t new_size) {
+            if(empty() || _is_self_allocated()) { // Valid for empty or self-allocated vector
+                _store.resize(new_size);
+                _data = _store.data();
+                _size = _store.size();
+            }
+            else { // raises error for mmap view
+                throw std::runtime_error("Cannot resize for mmap view case.");
+            }
+        }
+
+        void clear() { // Clear any storage
+            _size = 0;
+            _data = nullptr;
+            _store.clear();
+        }
+
+        /* Mmap save & load with MmapFile */
+        void save_mmap(MmapFile & mmap_f) const {
+            mmap_f.fput_one<uint64_t>(_size);
+            mmap_f.fput_multiple<T>(_data, _size);
+        }
+
+        void load_mmap(MmapFile & mmap_f) {
```

Comment:
similar name as `_load_mmap` but different concept, consider change to `from_mmap`? Similarly change `save_mmap` to `to_mmap`?